File: android/src/main/java/com/margelo/nitro/audioanvil/WavSegmentWriter.kt
package com.margelo.nitro.audioanvil

import java.io.File
import java.io.IOException
import java.io.RandomAccessFile

/**
 * One open WAV segment. Appends PCM, patches the header and fsyncs on an interval,
 * so the file on disk is a valid WAV at every flush point.
 */
internal class WavSegmentWriter(
  val file: File,
  val index: Int,
  val sampleRate: Int,
  val mediaStartMs: Double,
  fsyncIntervalMs: Double,
) {
  val startedAt: Double = System.currentTimeMillis().toDouble()
  var wasInterrupted = false
  var interruptionReason: InterruptionReason? = null
  var routeChanged = false

  private val output = RandomAccessFile(file, "rw")
  private val fsyncEveryBytes = maxOf(2, (sampleRate * 2 / 1000.0 * fsyncIntervalMs).toInt())
  private var bytesSinceFlush = 0

  var dataBytes = 0
    private set

  init {
    output.setLength(0)
    output.write(WavHeader.bytes(sampleRate, 0))
    output.fd.sync()
  }

  val durationMs: Double
    get() = dataBytes / (sampleRate * 2.0) * 1000.0

  fun append(bytes: ByteArray, count: Int) {
    output.seek(output.length())
    output.write(bytes, 0, count)
    dataBytes += count
    bytesSinceFlush += count
    if (bytesSinceFlush >= fsyncEveryBytes) {
      flush()
    }
  }

  /** Patches the header, then fsyncs. Header and data are consistent on disk afterwards. */
  fun flush() {
    WavHeader.patch(output, dataBytes)
    output.fd.sync()
    bytesSinceFlush = 0
  }

  fun complete(): RecordingSegment {
    flush()
    output.close()
    return RecordingSegment(
      index = index.toDouble(),
      filePath = file.absolutePath,
      sampleRate = sampleRate.toDouble(),
      durationMs = durationMs,
      fileSize = (WavHeader.BYTE_COUNT + dataBytes).toDouble(),
      mediaStartMs = mediaStartMs,
      startedAt = startedAt,
      endedAt = System.currentTimeMillis().toDouble(),
      wasInterrupted = wasInterrupted,
      interruptionReason = interruptionReason,
      routeChanged = routeChanged,
      sha256 = file.sha256Hex(),
    )
  }

  /** Best-effort close when `complete()` itself failed. The file stays valid up to the last flush. */
  fun abandon() {
    try {
      flush()
    } catch (_: IOException) {
    }
    try {
      output.close()
    } catch (_: IOException) {
    }
  }
}

File: ios/WavSegmentWriter.swift
import Foundation

/// One open WAV segment. Appends PCM, patches the header and fsyncs on an interval,
/// so the file on disk is a valid WAV at every flush point.
final class WavSegmentWriter {
  let index: Int
  let url: URL
  let sampleRate: Int
  let mediaStartMs: Double
  let startedAt: Double
  var wasInterrupted = false
  var interruptionReason: InterruptionReason?
  var routeChanged = false

  private let handle: FileHandle
  private let fsyncEveryBytes: Int
  private(set) var dataBytes = 0
  private var bytesSinceFlush = 0

  init(url: URL, index: Int, sampleRate: Int, mediaStartMs: Double, fsyncIntervalMs: Double) throws {
    self.url = url
    self.index = index
    self.sampleRate = sampleRate
    self.mediaStartMs = mediaStartMs
    self.startedAt = Date().timeIntervalSince1970 * 1000
    let bytesPerMs = Double(sampleRate * 2) / 1000.0
    self.fsyncEveryBytes = max(2, Int(bytesPerMs * fsyncIntervalMs))
    guard FileManager.default.createFile(atPath: url.path, contents: nil) else {
      throw AnvilError(.io, "Cannot create \(url.path)")
    }
    handle = try FileHandle(forWritingTo: url)
    try handle.write(contentsOf: WavHeader.bytes(sampleRate: sampleRate, dataBytes: 0))
    try handle.synchronize()
  }

  var durationMs: Double {
    Double(dataBytes) / Double(sampleRate * 2) * 1000.0
  }

  func append(_ pcm: Data) throws {
    _ = try handle.seekToEnd()
    try handle.write(contentsOf: pcm)
    dataBytes += pcm.count
    bytesSinceFlush += pcm.count
    if bytesSinceFlush >= fsyncEveryBytes {
      try flush()
    }
  }

  /// Patches the header, then fsyncs. Header and data are consistent on disk afterwards.
  func flush() throws {
    try WavHeader.patch(handle: handle, dataBytes: dataBytes)
    try handle.synchronize()
    bytesSinceFlush = 0
  }

  func finalize() throws -> RecordingSegment {
    try flush()
    try handle.close()
    let sha = try url.anvilSHA256Hex()
    return RecordingSegment(
      index: Double(index),
      filePath: url.path,
      sampleRate: Double(sampleRate),
      durationMs: durationMs,
      fileSize: Double(WavHeader.byteCount + dataBytes),
      mediaStartMs: mediaStartMs,
      startedAt: startedAt,
      endedAt: Date().timeIntervalSince1970 * 1000,
      wasInterrupted: wasInterrupted,
      interruptionReason: interruptionReason,
      routeChanged: routeChanged,
      sha256: sha
    )
  }

  /// Best-effort close when `finalize()` itself failed. The file stays valid up to the last flush.
  func abandon() {
    try? flush()
    try? handle.close()
  }
}
